src/services.py

```diff
@@ -0,0 +1,1033 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Define DataHub services."""
+
+# Some General Notes
+# 1. `charm._state` does not behave in the standard Python way, you cannot get a reference
+#    to an object inside it and update it. For lasting updates you need to an assignment with
+#    the updated object.
+# 2. `process.wait()` causes some scripts (i.e. datahub-upgrade) to hang, which is why we use
+#    `process.wait_output()` even though we do not read the output. It does not have the same
+#    problem for some reason.
+# 3. Flags in `charm._state` have three states: True, False, None. Logic for them is very specific
+#    depending on the point in the program lifecycle and at certain places, explicit checks with
+#   `is` is required over general truthiness checks.
+
+import abc
+import logging
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Dict, Optional
+
+import exceptions
+import literals
+import utils
+
+if TYPE_CHECKING:
+    from charm import DatahubK8SOperatorCharm  # noqa
+
+logger = logging.getLogger(__name__)
+
+
+def _kafka_topic_names(prefix: str) -> Optional[Dict[str, str]]:
+    """Compiles the environment variables for Kafka topic names.
+
+    Args:
+        prefix: Prefix to use for the topic names.
+    """
+    # Ref: https://github.com/datahub-project/datahub/blob/master/docs/how/kafka-config.md#topic-configuration
+    # These will be shared across all services unless a conflict is detected.
+    # In that case, each service will have environment variables strictly for itself.
+    default_names = {
+        "METADATA_CHANGE_PROPOSAL_TOPIC_NAME": "MetadataChangeProposal_v1",
+        "FAILED_METADATA_CHANGE_PROPOSAL_TOPIC_NAME": "FailedMetadataChangeProposal_v1",
+        "METADATA_CHANGE_LOG_VERSIONED_TOPIC_NAME": "MetadataChangeLog_Versioned_v1",
+        "METADATA_CHANGE_LOG_TIMESERIES_TOPIC_NAME": "MetadataChangeLog_Timeseries_v1",
+        "PLATFORM_EVENT_TOPIC_NAME": "PlatformEvent_v1",
+        "DATAHUB_UPGRADE_HISTORY_TOPIC_NAME": "DataHubUpgradeHistory_v1",
+        "DATAHUB_USAGE_EVENT_NAME": "DataHubUsageEvent_v1",
+    }
+    default_names["DATAHUB_TRACKING_TOPIC"] = default_names["DATAHUB_USAGE_EVENT_NAME"]
+
+    if prefix == "":
+        return default_names
+
+    names = {k: f"{prefix}_{v}" for (k, v) in default_names.items()}
+    return names
+
+
+@dataclass
+class ServiceContext:
+    """Context manager for service methods."""
+
+    # We can use the charm object straight up instead of having a 'ServiceContext' class.
+    # But, the first implementation used a 'ServiceContext' and I would like to leave an opening for future changes.
+    # TODO (mertalpt): Remove 'noqa' with the CI/CD update.
+    charm: "DatahubK8sOperatorCharm"  # noqa: F821
+
+
+class AbstractService(abc.ABC):
```

Comment:
[Suggestion] If this class is an abstract class, I'd suggest forcing implementation in subclasses. For this, you can use the `@abc.abstractmethod` decorator for functions inside it and throw a `NotImplementedError` exception. This will make sure that child classes implement functions accordingly. 

   ```
    @abc.abstractmethod
    def is_enabled(context: ServiceContext) -> bool:
        """Determine if the service should be enabled using configs and context clues.

        Args:
            context: Context for the service.
        """
        raise NotImplementedError
```

Reply:
Not all of the subclasses need to implement all the functions, I left them with good defaults instead.